{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HIV Time Since Infection Modelling\n",
    "Peter Bork"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We seek a method to determine the time since HIV infection for a given HIV-positive person using only Next Generation Sequencing (NGS) reads of the virus in that patient's blood. This document will derive a mathematical model from the concrete biological circumstances and apply it in code to solve this problem."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mathematical model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Basic setup\n",
    "Consider a set of locations in the genome of the first virus to infect a given person, $S$. $S$ can be the *pol* gene, the whole genome or, as will be important later, a selection of specific locations. The number of adenine bases will change from parent to offspring virus according to\n",
    "\n",
    "$$\n",
    "A_{g+1} = q_{AA} \\; A_g + q_{TA} \\; T_g + q_{CA} \\; C_g + q_{GA} \\; G_g - q_{AT} \\; A_g - q_{AC} \\; A_g - q_{AG} \\; A_g,\n",
    "$$\n",
    "\n",
    "where A, T, C and G refer to the four nucleotides and the subscript to the generation number of the virus ($g=0$ for the first virus entering the patient). The $q$ coefficients are the percent of nucleotides in the set which will be substituted in one generation, with $q_{AT}$ refering to the percent of adenines which become thymine (per generation) and so on."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Frequencies\n",
    "Divide through on both sides with the number of nucleotides in the set $S$ to get the frequency of nucleotides in the set rather than the absolute number:\n",
    "\n",
    "$$\n",
    "p(A_{g+1}) = q_{AA} \\; p(A_g) + q_{TA} \\; p(T_g) + q_{CA} \\; p(C_g) + q_{GA} \\; p(G_g) - q_{AT} \\; p(A_g) - q_{AC} \\; p(A_g) - q_{AG} \\; p(A_g),\n",
    "$$\n",
    "\n",
    "where we use the notation p(A) to refer to the frequency of the nucleotide and observe that we can similarly interpret this as the frequentist probability that a randomly chosen nucleotide in the set $S$ is adenine."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Matrix notation\n",
    "The equation for the frequency of adenines has natural parallels in equations for the three other bases, and we can conveniently arrange them using vector notation into\n",
    "\n",
    "$$\n",
    "F = [p(A), \\; p(T), \\; p(C), \\; p(G)]^T\n",
    "$$\n",
    "\n",
    "The difference equation can now be written as simply\n",
    "\n",
    "$$\n",
    "F_{g+1} = Q \\, F\n",
    "$$\n",
    "\n",
    "The matrix Q is a 4 by 4 matrix, where each column must sum to 1 because the sum of substitutions on a given set of bases must equal all the bases (since we count \"substitution\" to oneself). It therefore contains $4 \\times 4 - 4 = 12$ parameters.\n",
    "\n",
    "In the literature, Felsenstein 2004 calls this the general model (#REF)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### From difference to differential method\n",
    "(We'll use Newtons dot-notation to refer to differentiation w.r.t. generations rather than time)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Solution to the HIV Time Since Infection problem\n",
    "The difference equations above have analytical solutions. For the matrix version, the solution is\n",
    "\n",
    "$$\n",
    "F_g = e^{Q g} F_0\n",
    "$$\n",
    "(Where subscript g refers to generation number g).\n",
    "\n",
    "\n",
    "Using knowledge of the substitution rates, initial set of nucleotides ($S_0$) and final frequency of nucleotides in the corresponding current set ($S_g$), the model can therefore theoretically be solved for the number of generations which have passed since infection."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Fundamental problem\n",
    "The fundamental problem with the model is that the solution in the form presented above requires knowledge about 12 parameters which are not all easy to determine.\n",
    "\n",
    "A second, smaller problem, is that the model requires knowledge of the initial infecting set $S_0$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Model Assumptions\n",
    "This formulation disregards the stochastic nature of substitutions, but over many generations and many vira, this assumption should be permitted.\n",
    "\n",
    "The linear nature of this formulation also assumings that the number of bases in the set are near equilibrium such that we do not see non-linear effects (if e.g. the ratio of thymines to other bases were three times larger than freely available nucleotides, we would expect the substitution away from thymine to be greater in proportion to this discrepancy until equilibrium had been reached). (Using Hartman-Grobman we can show that the linear behavior should approximate the behavior of the non-linear system near the equilibirum)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using biological insight to reduce the problem\n",
    "\n",
    "Biological insight allows us two crucial ways to apply the model presented above:\n",
    " 1. By carefully choosing the set of nucleotides we can reduce the dynamical system. \n",
    " 2. Some substitution rates can be estimated from NGS data due to ther potential for \"death codons\" (see below)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Death-reduced genome locations\n",
    "Of the 64 codons available to the protein-coding genes, TAA, TAG are two stop codons with only a single letter substitution between them in the third position. \n",
    "\n",
    "In currently living virus populations, TAT and TAC codons in the middle of crucial protein-coding genes cannot ever have had an A or a G in the third position, since the population would then have ended at that generation. The nucleotide alphabet in these positions are therefore reduced from four to two nucleotides as long as the two other nucleotides remain TA.\n",
    "\n",
    "A change in either of the first two nucleotides will result in a change in amino acid away from tyrosine. Since tyrosine is the rarest of amino acids in the codon table - being represented only by TAT and TAC, - we should expect evolutionary fitness costs to weigh against the TAT and TAC codons changing much.\n",
    "\n",
    "The set of locations in the genome which are the third base in TAT or TAC codons are $S_{RD}$, the set of reduced death."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Death-reduced nucleotide evolution has reduced model complexity\n",
    "The mathematical model above reflects the frequency of nucloetides in a freely chosen set of nucleotides $S$. Let us choose only those positions in the set of reduced death, $S_{RD}$.\n",
    "\n",
    "It is crucial to remember that the model reflectes the history of a currently \"living\" virus population under the assumption that this population has maintained the tyrosine-coding codons as such throughout its history. Because the number of adenine and guanine nucloetides in the third position of the TAC and TAT codons cannot have happened,\n",
    "\n",
    "$$\n",
    "p(A) = p(G) = 0 \\quad for \\quad A,G \\in S_{RD}\n",
    "$$\n",
    "\n",
    "We know also that all rates of substitution into adenine and guanine in the history of a chosen virus will be zero in $S_{RD}$. Similarly, all rates of substitution from adenine or guanine to the other four nucleotides will have been zero in the history of the \"living\" virus."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Reduced mathematical model\n",
    "Applying these insights in the model derived above yields a simple system of two couple differential equations,\n",
    "\n",
    "$$\n",
    "\\dot{p}(C) = q_{TC} \\; p(T) - q_{CT} \\; p(C) \\quad for \\quad C, T \\in S_{RD} \\\\\n",
    "\\dot{p}(T) = q_{CT} \\; p(C) - q_{TC} \\; p(T) \\quad for \\quad C, T \\in S_{RD}\n",
    "$$\n",
    "\n",
    "Since $p(C) + p(T) = 1$, we can reduce this system to a single differential equation\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "\\dot{p}(C) &= q_{TC} \\; (1 - p(C)) - q_{CT} \\; p(C) &\\quad for \\quad C \\in S_{RD} \\\\\n",
    "           &= q_{TC} - (q_{TC} + g_{CT}) \\; p(C)    &\\quad for \\quad C \\in S_{RD}\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "Only two parameters need now be determined, $q_{TC}$, and $q_{CT}$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Solution to reduced model\n",
    "\n",
    "The solution is naturally the sum of the inhomogenous and the homogenous differential equations. At equilibrium, where $\\dot{p}(C) = 0$, we find by substitution the expected frequency of cytosines to be\n",
    "\n",
    "$$\n",
    "\\begin{equation}\n",
    "p^*(C) = \\frac{q_{TC}}{q_{TC} + q_{CT}} \\quad for \\quad C \\in S_{RD}\n",
    "\\end{equation}\n",
    "$$\n",
    "\n",
    "The fundamental solution is therefore\n",
    "\n",
    "$$\n",
    "p_g(C) = \\frac{q_{TC}}{q_{TC} + q_{CT}} + p_0 (C) \\;  e^{-(q_{TC} + g_{CT}) g} \\quad for \\quad C \\in S_{RD}\n",
    "$$\n",
    "\n",
    "Given the starting and final frequencies of cytosine, we can therefore estimate the number of generations of infection using\n",
    "\n",
    "$$\n",
    "\\hat{g} = \\frac{-1}{q_{TC} + q_{CT}} \\; \\ln\\Bigl(\\frac{p_g(C)}{p_0(C)} - \\frac{Q_{TC}}{p_0(C) \\, (q_{TC} + q_{CT})}\\Bigr) \\quad for \\quad C \\in S_{RD}\n",
    "$$\n",
    "\n",
    "As is evident from the logarithmic term, we require\n",
    "$$\n",
    "\\frac{p_g(C)}{p_0(C)} > \\frac{Q_{TC}}{p_0(C) \\, (q_{TC} + q_{CT})} \\quad\\Rightarrow\\quad\n",
    "p_g(C) > \\frac{q_{TC}}{q_{TC} + q_{CT}} \\quad\\Rightarrow\\quad\n",
    "p_g(C) > p^*(C) \\quad for \\quad C \\in S_{RD}\n",
    "$$\n",
    "\n",
    "In other words, the observed frequency of cytosines must be greater than the equilibrium level."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Final generation death rate in NGS data\n",
    "Even though the probability of a death codon having occurred in the history of the extant virus is zero, we will still find some death codons in NGS data since the last generation will sometimes get such substitutions. The solutions above therefore reflect the frequencies up until the generation preceeding the final generation.\n",
    "\n",
    "When calculating the frequencies of thymine substitutions for $TAC \\rightarrow TAT$ codons, we can consider all guanines and adenines in the third position to be outside of the virus population we are modelling. In NGS data, our $p(T)$ in these positions should therefore be calculated as\n",
    "$$\n",
    "p(C) = \\frac{p(C)}{p(C) + p(T)}  \\quad for \\quad C, T \\in S_{RD}\n",
    "$$\n",
    "rather than\n",
    "$$\n",
    "p(T) = \\frac{p(T)}{p(C) + p(T) + p(G) + p(A)}  \\quad for \\quad A, T, C, G \\in S\n",
    "$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Estimating the two substitution rates\n",
    "We wish to estimate the rate of substitution from thymine to cytosine and vice versa. We can use two lines of arguments,\n",
    " 1. The proportion of vira with death codons in which a C has been swithces to T represent the frequency of C to T substitutions\n",
    " 2. The observed overall ratio of TAC to TAT must be the equilibrium of the two rates and only one of these need therefore be measured."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Estimating substitution rates via death codon\n",
    "The codons CGA, CAA and CAG can become a stop codon with a single substitution of $C \\rightarrow T$. The frequency of these substitutions within crucial protein-coding genes correspond therefore to the substitution rate $q_{CT}$.\n",
    "\n",
    "The set of locations which are first nucleotide in CGA, CAA or CAG codons in the first 95% of a crucial protein-coding gene is $S_{DCC}$, the set of death codon cytosines.\n",
    "\n",
    "Unfortunately, the substitution rate $T \\rightarrow C$ cannot be determined by a similar measurement. Fortunately, we observe the ratio of TAT codons to TAC codons which must be determined by the rates $q_{CT}$ and $q_{TC}$.\n",
    "\n",
    "The equilibrium solution for the frequency of cytosines in set $S$ is mirrored by a similar solution to the frequency of thymines. The ratio of these two depend only on the two substitution rates we seek, and since we know both the ratio and one of the rates, we can simply solve for the other.\n",
    "\n",
    "Use $p^*(T) = (1-p^*(C)$ and a bit of algebraic manipulation to find\n",
    "$$\n",
    "q_{CT} = \\frac{p^*(C)}{p^*(T)} \\; q_{TC}  \\quad for \\quad C, T \\in S_{RD}\n",
    "$$\n",
    "\n",
    "Assuming $S_0$ has the equilibrium ratio of cytosines to thymines we can now estimate both substitution rates."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Estimating $p_0(C)$ and $p_g(C)$\n",
    "For bloodsamples where the TAT and TAC codons have no reads with the opposite nucleotide in the third position, we must assume that the virus has simply not substituted these locations. In fact, as long as the frequency of the alternate base is relatively far from equilibrium, we can identify the original virus' base as the most frequent one.\n",
    "\n",
    "For samples which have sufficiently short infection time, we are therefore able to determine both $p_g(C)$ and $p_0(C)$."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Coding the solution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We shall construct a python program which can eat a directory of pileup files and output a tab-separated file with infection age predictions as well as diagnostics for each pileup.\n",
    "\n",
    "The program may proceeed for each pileup file to\n",
    " 1. Read the *pol* gene and identify $S_{RD}$ and $S_{DCC}$ positions based on majority frequencies\n",
    "    - Find start-codon, use minimum frequency of all three codon bases.\n",
    " 2. Set $q_{CT}$ equal to the rate of death codons based on CGA, CAA and CAG codons having frequencies for TGA, TAA and TAG.\n",
    "    - Do not use codons near the 5% end of the pol gene.\n",
    " 3. Count the number of TAT and TAC codons based again on majority frequency (and frequency cut-off), and set $q_{TC}$ equal to $\\frac{TAC}{TAT} \\, q_{TC}$, (as per above).\n",
    " 4. For each TAT codon determined by majority frequency cut-off, find $p(C)$ and calculate $\\hat{g}$\n",
    " 5. For each TAC codon determined by majority frequency cut-off, find $p(T)$ and calculate $\\hat{g}$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Estimate time since infection ($\\hat{g}$)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from typing import List"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimate_generations(pgs: float or List[float], qi: float, qo: float,\n",
    "                         p0: float=None) -> float:\n",
    "    \"\"\" Estimates number of generations from zero to mutation_rate_g\n",
    "    \n",
    "    Assumes positions are TAC or TAT position, and should be inside crucial protein-\n",
    "    coding gene (not nearer than 5% to the end of the gene).\n",
    "    \n",
    "    Assumes initial condition p0 is equilibrium frequencies (qi / (qi + qo)) if not given.\n",
    "    \n",
    "    based on\n",
    "    g = -(1 / (qi + qo)) * ln[(pg / p0) - qi / (p0*(qi + qo))],\n",
    "    where\n",
    "    :param pgs: mutation percentage at generation g of major base for each position\n",
    "    :param qi: rate of mutations into major base\n",
    "    :param qo: rate of mutations away from major base\n",
    "    :param p0: initial ratio of major base to minor base (cytosine to thymine)\n",
    "    :return: float estimated generation number between original virus and current.\n",
    "    :raises ValueError: if pg is less than equilibrium level.\n",
    "    \"\"\"\n",
    "    pgs = [pgs] if type(pgs) is float else pgs\n",
    "    p_equilibrium = qi / (qi + qo)\n",
    "    p0 = p0 or p_equilibrium\n",
    "    log_args = [(pg/p0) - qi / (p0*(qi + qo)) for pg in pgs]\n",
    "    \n",
    "    if any([log_arg <= 0 for log_arg in log_args]):\n",
    "        pg_min = min(pgs)\n",
    "        raise ValueError('Estimating generation number with mutation frequency less '\n",
    "                         'than equilibrium is not possible.\\n Minimum mutation frequency given '\n",
    "                         'is pg=%.2f, whereas equilibrium is %.2f.' % (pg_min, p_equilibrium))\n",
    "\n",
    "    return -(1 / (qi + qo)) * np.log(log_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assuming 1.2 days per generation, we get the following relation between substituion frequency and time since infection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAAEYCAYAAAAJeGK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XecVOXZ//HPtY3eQaR3CyrdihBLbNGoSdRoxF6fGKMm\nxhifPI8m0Wg0ifHRWLCCEhU79i6KKNKVJr0jSGeBZRnm+v1xzuLIb5edXWb2zMx+36/XvnZOv84B\n5uIu577N3REREck0eVEHICIiUh4lKBERyUhKUCIikpGUoEREJCMpQYmISEZSghIRkYykBCUZw8xu\nMrNHIo7hTTO7oIavWWxmXZPcd6CZzQmPOT2FMXQMz5mfqnOK7CnTe1BSU8ysOGGxPrAN2BEuX+Hu\nI2o+quxiZu8Do9z9nj08z0LgUnd/LyWBiaRBQdQBSO3h7g3LPusLsto6AdOjDkKkJqiKTzKGmd1i\nZk+FnzubmZvZRWa2xMzWmdmVZnawmX1pZuvN7L5djr/YzGaG+75tZp0quE5dM3vKzNaE5xlvZq3D\nbR+Z2aXh5wvNbIyZ/T085wIzOynhPM3N7HEzWx5ufzlh2ylmNiU8/1gz67Wb+3Yz6x5+fsLM/m1m\nr5vZJjMbZ2bdwm3zgK7Aq2F1XB0za2Jmj5rZCjNbZma3JlbTmdll4TPZZGYzzKyfmT0JdEw4zw0J\nz7sgPK6tmY0ys7VmNtfMLtvlz2mkmQ0PzzvdzAYk++cskiwlKMl0hwI9gJ8D/wL+G/ghcABwlpn9\nAMDMTgNuAn4KtAI+AZ6u4JwXAE2ADkAL4Epg626u/zXQErgTeNTMLNz2JEFV5QHAXsDdYSx9gceA\nK8LzPwSMMrM6Sd7z2cCfgGbAXOA2AHfvBiwGfuzuDd19G/AEEAO6A32B44GyBHsmcAtwPtAYOBVY\n4+7n7XKeO8uJ4RlgKdAWOAP4q5kdk7D91HCfpsAo4L7/7wwie0gJSjLdX9y9xN3fATYDT7v7Kndf\nRpCE+ob7XQnc7u4z3T0G/BXoU0EpajtB4uju7jvcfaK7b6zg+ovc/WF33wEMA9oArc2sDXAScKW7\nr3P37e4+OjzmcuAhdx8Xnn8YQXvbYUne80vu/kV4HyOAPuXtFJb6fgRc6+6b3X0VQZI8O9zlUuBO\ndx/vgbnuvqiyi5tZB2Ag8Pvw2U8BHiFIdGXGuPsb4XN5Euid5L2JJE0JSjLdyoTPW8tZLmvX6gTc\nE1aprQfWAga0K+ecTwJvA8+E1XN3mllhBdf/puyDu28JPzYkKH2tdfd15RzTCfhtWSxhPB0ISiPJ\n+Cbh8xa+u8fyrlMIrEi4zkMEpTnCa85L8pqJ2hLc26aEdYv4/rPcNca6ZdWDIqmiv1CSK5YAtyXT\nE9DdtxNUof3JzDoDbxBU4z1axes1N7Om7r6+glhuq8L5qmMJQcmsZVjaKm97twqO3V333eUE99Yo\nIUl1BJZVO1KRalAJSnLFg8AfzOwAgLDzwJnl7WhmR5vZQWFngo0EVX7xqlzM3VcAbwL3m1kzMys0\ns8Hh5oeBK83sUAs0MLOTzaxRdW9uNzG8A/zDzBqbWZ6ZdStrlyOolrvezPqHcXRPqPJcSdDhorzz\nLgHGAreHHUp6AZcAT6UyfpHKKEFJTnD3l4C/EVTbbQSmEbQRlWdv4HmC5DQTGE1Q7VdV5xEkt1nA\nKuDaMJYJwGUEHQfWEXR0uLAa50/G+UARMCO81vME7WS4+3MEHSz+A2wCXgaah8fdDvwxrBq8vpzz\nngN0JihNvQTcrFcCpKbpRV0REclIKkGJiEhGUoISEZGMpAQlIiIZSQlKREQyUla8B5WXl+f16tWL\nOgwRkVphy5Yt7u6RF2CyIkHVq1ePzZs3Rx2GiEitYGYVjU1ZoyLPkCIiIuVRghIRkYykBCUiIhlJ\nCUpERDKSEpSIiGQkJSgREclISlAiIpKRlKBERHLIl0vX87vnprJhS2nUoewxJSgRkRzyv69M57mJ\nS5n1zabKd85wSlAiIjli1cYSpixZz96N63Bo1xZRh7PHlKBERHLEn16dDsDVx/aIOJLUUIISEckB\nsVict6evpH5RPucc3CHqcFJCCUpEJAfc/f4cYnHnrAEdyMvLja/23LgLEZFabvhnC8k348YT9406\nlJRRghIRyXIvT17GppIYP9i3JXWLsmIWpaQoQYmIZLl/vPM1AH867cCII0ktJSgRkSz25dL1LFm3\nlf3bNKJDs/pRh5NSSlAiIlnsllFB1/KbTzkg4khSTwlKRCRLrS4uYfLi9bRuVIfDumX/i7m7UoIS\nEclSt4yagQNXHdM96lDSQglKRCQLxWJx3pr2DfUK8xlyaMcavbaZ7WtmUxJ+NprZtam+Tu70RxQR\nqUX+Fb6Y+4tD29f4i7nu/jXQB8DM8oFlwEupvo5KUCIiWWjYZwvJM7jxpP2iDuVYYJ67L0r1iVWC\nEhHJMg9/Mo9NJTGO238v6qfnxdwCM5uQsDzU3YdWsO/ZwNNpCSIdJxURkfSIx+P867055BncdWbv\ndF0m5u4DKtvJzIqAU4E/pCMIVfGJiGSR//tgLpu37eCUXm1pWr8o6nBOAia5+8p0nFwJSkQkS8Tj\ncR4cPY/8POOvPzko6nAAziFN1XugBCUikjVuf3MWJdvjnNG/PQ3rRttCY2YNgOOAF9N1DbVBiYhk\ngdJYnCfGLqQw3/hLBgwK6+6bgbQOX6ESlIhIFrhl1DS273DOP7wzRQW146u7dtyliEgW21IaY+SE\npdQtyOMPJ0b+3lONSVuCMrMOZvahmc0ws+lmdk24vrmZvWtmc8LfzdIVg4hILrjpxWnE4s4VP+hK\nQS0pPUF6S1Ax4Lfu3hM4DLjKzHoCNwLvu3sP4P1wWUREyrFhSymjpi6jflE+1xzbI+pwalTaEpS7\nr3D3SeHnTcBMoB1wGjAs3G0YcHq6YhARyXY3PP8lcYfrftijxsfci1qN3K2ZdQb6AuOA1u6+Itz0\nDdC6JmIQEck2q4tLeGfmSprUK+Cywd2iDqfGpT1BmVlD4AXgWnffmLjN3R3wCo673MwmmNmEWCyW\n7jBFRDLOVSMm4w6/P3H/qEOJRFoTlJkVEiSnEe5e9jLXSjNrE25vA6wq71h3H+ruA9x9QEGBXtcS\nkdrl83lrGLdgLe2a1uMXNTzfU6ZIZy8+Ax4FZrr7PxM2jQIuCD9fALySrhhERLLV1c9MBuCBIf0i\njiQ66SyaDATOA74ysynhupuAO4CRZnYJsAg4K40xiIhknXven823m7Zx1L6t6NW+adThRCZtCcrd\nxwBWweZj03VdEZFstqlkO/e+P5fCfOO+c2pv6Qk0koSISEa56j+TiMWda47tEfmAsFFTghIRyRBT\nl6zj49mr2atRHX51TO16Kbc8SlAiIhniv56aBMC95/SNOJLMoAQlIpIBHv5kHss3lHBY1+Yc2jWt\ns1hkDSUoEZGIbSmNcddbs8nPMx4c0j/qcDKGEpSISMSufWYKpTviXDG4K03rF0UdTsZQghIRidAX\nC9bwzoyVNKtfyPXH7xN1OBlFCUpEJCLxeJzLhk8E4OHzB2TVaOVm1tTMnjezWWY208wOT/U1ancn\nexGRCF377FQ2bN3OaX3aMqBz86jDqap7gLfc/QwzKwLqp/oCSlAiIhH4YsEaRk1dTtN6hdx9Vu+o\nw6kSM2sCDAYuBHD3UqA01dfJnvKkiEiOiMXiXDZ8AgCPXJCRVXsFZdMdhT+X77K9C/At8LiZTTaz\nR8ysQaqDyLinIiKS6657bgobtsY4vU+7TK3ai5VNdxT+DN1lewHQD3jA3fsCm4EbUx2EEpSISA0a\nN38Nr05dQdP6hfzzrF5Rh1NdS4Gl7j4uXH6eIGGllBKUiEgNSazae/SCgzOxai8p7v4NsMTM9g1X\nHQvMSPV11ElCRKSGXDtyChtLYvy0bzv6d2oWdTh76mpgRNiDbz5wUaovoAQlIlIDxs5dzWtfrqBZ\n/UL+fmbWVu3t5O5TgAHpvEZ2li9FRLJIcUmMS4YFVXuPXZi9VXs1TU9JRCTNfv7QZ2zdvoMrBnel\nb8esr9qrMUpQIiJpdNfbXzN9xUb237sRf/jR/lGHk1WUoERE0mTionXc/+Fc6hbmMfLKlA9Vl/OU\noERE0mBLaYzzHx2HA4+cP4BGdQujDinrKEGJiKTBLx4ex+bSHVx4RGeO7NEq6nCykhKUiEiK3fP+\nbKYsWU/3vRpyy6kHRB1O1lKCEhFJoS+Xrudf786hTkEez6vdaY8k/aKumTUD2gJbgYXuHk9bVCIi\nWai4JMa5DwftTg8M6a/p2/fQbhNUOOfHVcA5QBHB8Op1gdZm9jlwv7t/mPYoRUSywOn3f8qmbTHO\nP7wTx+y3V9ThZL3KSlDPA8OBQe6+PnGDmfUHzjOzru7+aLoCFBHJBlc/PYm5q4rp27Epfz7twKjD\nyQm7TVDuftxutk0EJqY8IhGRLPPoJ/N5deoKWjQo4tnL1e6UKpW2QZmZAYcA7cJVy4Av3N3TGZiI\nSDYYN38Nt74+k8J847Wrj6SoQH3PUqWyNqjjgfuBOQSJCaA90N3Mfunu76Q5PhGRjLVqYwnnP/YF\nTjAIbJum9aIOKadUVoK6B/ihuy9MXGlmXYA3AA0sJSK1UiwW55R7x7AtFud3J+zLIL2Mm3KVlUUL\nCKb23dUyQON2iEitde6j41i1aRvH92zNVUd3jzqcnFRZCeoxYLyZPQMsCdd1AM4G1HNPRGql/31l\nGuMWrKVTi/o8OKRf1OHkLKusr4OZ9QRO5fudJEa5e8rnn69IgwYNfPPmzTV1ORGRCj00eh63vzmL\nRnULGHvjMTk5CKyZbXH3BpXssxDYBOwAYu6e8tl1K+3FFyaiGktGIiKZ6rWpy7n9zVkU5efxxjWD\ncjI5VdHR7r46XSffbRuUmTUxszvMbJaZrTWzNWY2M1zXtJJjHzOzVWY2LWHdLWa2zMymhD8/StWN\niIik0xcL1vDrZyaTZ/DclYfRoVn9qEPKeZV1khgJrAOOcvfm7t4CODpcN7KSY58ATixn/d3u3if8\neaOqAYuI1LQF3xZz7iPjiDs8OKQ/vTvk/LTtBWY2IeHn8nL2ceA9M5tYwfY9D6KS7Z3d/W/fi8j9\nG+BvZnbx7g5094/NrPOehSciEq21xaWccu8Ytu9wbv5xT44/YO+oQ6oJybQpHenuy8xsL+BdM5vl\n7h+nMojKSlCLzOwGM2tdtsLMWpvZ7/muV19VXW1mX4ZVgBX+N8TMLi/L3rFYrJqXEhGpvpLSGMf/\n62M2l+7g0kFduGhgl6hDyhjuviz8vQp4iWDEoZSqLEH9HGgBjA7boNYCHwHNgbOqcb0HgK5AH2AF\n8I+KdnT3oe4+wN0HFBQkPSuIiEhKxONxTr53DKuLt3HSgXvzx5N7Rh1SxjCzBmbWqOwzcDwwbfdH\nVV1lg8WuA34f/uwxd19Z9tnMHgZeS8V5RURSKR4PRomY9+1m+nVsygND+kcdUqZpDbwUDNVKAfAf\nd38r1RepdtHEzC5y98ereEwbd18RLv6ENGRcEZE9EY/H+ekDnzFjxSZ6tmmkWXHL4e7zgd7pvk6l\nL+pWeKDZYnfvuJvtTwNHAS2BlcDN4XIfgt4fC4ErEhJWhfSirojUlLOHfsbn89fSY6+GvH3tIPLy\nat/o5Mm8qFsjcewuQZnZlxVtAvZx9zppiWoXSlAiUhMueOwLRs/+lk4t6vP+dT+goJZOnZEpCaqy\nKr7WwAkE7z0lMmBsWiISEYnAFcMnMHr2t7RrWo93a3FyyiSVJajXgIbuPmXXDWb2UVoiEhGpYb9+\nejJvz1hJ60Z1eP83gzXpYIaodhtUTVIVn4ikyzXPTOaVKctp0aCI0b87moZ19VpLplTxVTYWX8PK\nTpDMPiIimeiyYeN3JqcPfvsDJacMU1k59hUz+4eZDQ5fxgLAzLqa2SVm9jblj7cnIpLRhjzyOe/O\nXMXejesw+ndH06R+UdQhyS4qe1H32HDE8SuAgeHQRDHga+B14IJwbD4RkawQj8c548HPmLR4PR2b\n1eOd6wZTt0glp0ykNigRqTXKhi+auWIT3Vo15O1rBqm3XjkypQ1K/20QkVqhNBbnpHs+Zt63mzmg\nTWNevXpgrXwJN5soQYlIzisuiXHiPR+zdN1WBnRqxsgrDlNyygJKUCKS01ZuLOGEuz9m/dbtDN6n\nJcMvPjTqkCRJSf0Xwsy6mVmd8PNRZvbryqZ8FxGJ2owVG/jBXR+yfut2zhrQXskpyyRbxn0B2GFm\n3YGhQAfgP2mLSkRkD43+ehU/vvdTSrbHue64Htx5RtoH35YUS7aKL+7uMTP7CXCvu99rZpPTGZiI\nSHU9O34xN77wFQB/P6MXZwzoEHFEAmDBBFL13T2pbtnJlqC2m9k5wAV8N8lgYTXiExFJq3++8zW/\nf+Er8vOMJy85RMkpYmY23Mwam1l94Ctgrpn9Jpljk01QFwGHA7e5+wIz6wI8Wb1wRUTS41f/mcT/\nfTCXeoX5vHb1kRzZo1XUIQn0cveNwOnAu0An4MJkDkyqis/dZwC/TlheAPytymGKiKRBSWmM0/49\nlq9XbqJlgyJev2YQrRvXjTqsnGdm+cAEYJm7n1LBboVmVgCcBjzg7qVmFk/m/EklKDMbCNxCkPkK\nCOaDcnfvmszxIiLpsmTdFn587xjWb9nOQe0a88J/DdR0GTXnGmAm0Hg3+zwMLAamAaPNrCNQnMzJ\nk+0k8ShwHTAR2JHkMSIiaTVmzrdc9MR4tu9wftavHf84q0/UIdUaZtYeOBm4DSi3TSksYS1x97YJ\n65YCxyRzjWQT1AZ3fzPJfUVE0u6xMQv4y2szAPifk/fnkkGq0EmhAjObkLA81N2H7rLPv4AbgEYV\nncTdd5jZTQSvKpWtiwOlSQWRZLAfmtldwIvAtoQLTUryeBGRlPnNs1N4cfIyCvONxy88WJ0hUi/m\n7gMq2mhmpwCr3H2imR1VybneMbNrgWeBnd3Lw44Tu5XUaOZm9mE5q93dkyqm7SmNZi4iAJtKtnP6\nv8cy79timtYv5NWrj6RDs/pRh5VzKhvN3MxuB84jmH6pLkEb1IvuPqScfZeUcwp3946VxqHpNkQk\nG0xdso5zHh7HltId9O3YlGcvP1ydIdKkKtNthCWo63fTi6/aku3F1wS4GRgcrhoN/NndN6Q6IBGR\nXT36yXxufX0mDlwxuAt/+FHPqEOSKjCz/YCeBKUtANy90uHykq3ie4Ggi+CwcNV5QG93/2m1oq0i\nlaBEaqd4PM4VT07k3ZmrKMw3HhzSn2P3bx11WDkvlRMWmtkfgeOB/YC3gROAMcnkj2QT1BR371PZ\nunRRghKpfVZtLOH0+8eyfP1W9mpUh1euGkibpvWiDqtWSHGC+groA0xy995m1gZ4wt1PqOzYZHvx\nbTWzI919THjBgcDWakcsIrIbr01dznUjp7B9hzOoR0uGXXSwJhjMXlvD7uYxM2sEfEMw6EOlkk1Q\n/wUMC9uiDFhLkmMpiYgkKx6Pc93IqbwyZTl5pvebcsTkcP7AxwiGRdoIfJHMgVXqxWdmjSG5/uup\npCo+kdy3Yv1WfvbAWJZvKKFp/UJGXnE4+7Su8B1QSaNUVvHtct7uQONk36HdbQnKzIa4+1O7Do0e\nTOkB7v7P6gYqIlLm5cnLuP65qcTizuB9WvLY+QdToC7kOcPMzga6ufttZtbBzPq7+8TKjqusiq8s\ng5b335jMf4FKRDJaPB7nlyMm8db0leQZ/Pm0Azj/8M5RhyUpZGb3EcwfOJhg3L7NwIPAwZUdu9sE\n5e4PhR/fc/dPd7nowGpFKyICzFi+gXMfGce6Ldtp2aCI5648nC6tGkYdlqTeEe7er2wWdndfa2ZF\nyRyYbCeJe4F+SawTEanU396cxYOj5+HAj3u34Z6f91Evvdy13czyCGvdzKwFsOfzQZnZ4cARQKtd\n2qEaA/nVi1VEaqtVG0s4e+jnzF+9mbqFedx/bn+O2W+vqMOS9Po3wWjmrczsT8BZwJ+SObCyElQR\n0DDcL7EdaiNwRtXjFJHaauSEJdz04lfE4k7fjk158uJDaVg32UocyTZm9gbwS3cfbmYTgR8SvKZ0\nprtPS+ocSY4k0cndF1UxuMeAsiHZDwzXNScYcr0zsBA4y93XVXYudTMXyV7FJTEueuILxi9cR77B\nf5/ck4uP7BJ1WLIbqehmbmZnEnSKGAbc6e7bq3yOJBNUK4KJqQ7g+4P9VTjdhpkNJpjWd3hCgroT\nWOvud5jZjUAzd/99ZddXghLJTq9MWcYNz3/Jtlicjs3rMeKywzQ9RhZI1XtQZtYQ+B/gROBJEtqe\nknlNKdny9QiCks8pwJXABcC3uzvA3T82s867rD4NOCr8PAz4CKg0QYlIdikuiXHJsPGMW7CWPINf\nHd2d60/YN+qwpOaVEnQrr0PQTJRU54gyySaoFu7+qJld4+6jgdFmNr5qcQLQ2t1XhJ+/ASocltjM\nLgcuBygqSqpHoohkgNemLuf656ZSEovToVk9nrr0UDq1SPmgBJLhzOxE4J/AKKCfu2+p6jmSTVBl\ndYcrzOxkYDnQvKoXS+TubmYV1i+6+1BgKARVfHtyLRFJv00l27ls+AQ+n78WM/jlUd244cT9og5L\novPfBB0iplf3BMkmqFvDgWJ/S/D+U2Pgumpcb6WZtXH3FeGQ66uqcQ4RyTAjxi3illHT2b7Dad+s\nHk9efIheuq3l3H3Qnp6j0gRlZvlAD3d/DdgAHL0H1xtF0H51R/j7lT04l4hEbMX6rVzw+BfMXllM\nvhnXHdeDa47dJ+qwJEck24vvC3c/pEonNnuaoENES2AlwZTxLwMjgY7AIoJu5msrO5d68Ylknrve\n/poHPppL3KFnm0Y8cdEh7NW4buUHSsZL12jmVY4jyQR1N8Fgf88S9MgAINkh0/eUEpRI5vhy6Xou\nHz6BbzZuo05BHn8+7QB+fnDHqMOSFKosQZlZXeBjgt55BcDz7n5zyuNIMkF9WM5q3917UKmkBCUS\nvZLSGFc/PZl3ZwZNx0d2b8HQ8wdQv0ijQeSaJBKUAQ3cvdjMCoExwDXu/nkq40jqb5a770m7k4hk\nuac+W8hfXp/Jtlic5g2KuO+cvhzRvWXUYUlEPCjZFIeLheFPyntbJ5WgzKw18FegrbufZGY9gcPd\n/dFUByQimWPet8VcOmw8C1ZvId/gisFd+f2J+2rk8dxXYGYTEpaHhq/+7BR2oJsIdAf+7e7jUh1E\nslV8bwKPA//t7r3NrACY7O4HpTqg8qiKT6Rmlcbi/O75qYyashwHendowsPnDVAniFqiKp0kzKwp\n8BJwdbKDwCYr2crjlu4+0sz+AODuMTPbkcpARCQzDBu7kNvfmElJLE6jugX848zeHH/A3lGHJRnK\n3deH/RROBCJJUJvDSabKJpw6jOCdKBHJEZMXr+OqEZNYvqGEfIOLB3bmjyfvr+o8+f+EA4hvD5NT\nPeA44G+pvk6yCeo3BC/ZdjOzT4FWaD4okZywfksp//XUJD6bvwaAQ7s054Fz+9O8ocbAlAq1AYaF\n7VB5wMhwMIeUSqoNCiBsd9qXYMKpr6szt0d1qQ1KJPXi8Ti3vDqDEZ8vZoc7bZrU5d5z+jKg8x4N\nsyk5IFNe1K3KCwyHEEw0WAD0MzPcfXhaohKRtHrqs4X89c1ZbCndQZ2CPG48YV8uG9Qt6rBEvifZ\nbuZPAt2AKUBZ5wgHlKBEssjYuau5buQUVm7cRp7BWQPac+vpB1FUoHYmyTzJlqAGAD092fpAEcko\ni9Zs5pcjJjF9+UYADu/agnt/0YeWDdVtXDJXsglqGrA3sKKyHUUkc6wtLuXakZP5ePZqALq0bMC/\nf9GXnm2bRByZSOWSfg8KmGFmXwDbyla6+6lpiUpE9khJaYwbX5zGqKnLiDu0aFDErT85kJMObBN1\naCJJSzZB3ZLOIEQkNeLxOLe/OYsnxi5k+w6nQVE+N5y4Hxcc0Tnq0ESqLNnBYkebWSeCiQvfM7P6\nQH56QxORZMXjcR78eD73fTCXLaU7KMrP45dHdeX64/fRi7aStZLtxXcZcDnQnKA3XzvgQeDY9IUm\nIsl4/NMF/PPd2WwqiZFvxpn923PbT9QzT7JfslV8VxG8BzUOwN3nmNleaYtKRCo1csIS/vrGTNZv\n2U6ewckH7c0dP+tFo7qFUYcmkhLJJqht7l4azFG1c1QJdTkXicCoqcv486szWV28DQOO3W8v7jqj\nt4YmkpyTbIIabWY3AfXM7Djgl8Cr6QtLRHb18uRl3Pp6kJggmNH2H2f1obWmwJAclex8UHnAJcDx\nBGPxvQ08UlMv7mosPqnNXpq0jNvemMHq4lIgeMn2zjN70aFZ/Ygjk1yVKWPxJT1YbJSUoKQ2Gjlh\nCX97cxZrNpdiwBHdgsTUrqkSk6RXpiSo3VbxmdlpQHt3/3e4PI5gqg2A37v7c2mOT6RWicfjPP7p\nQu79cC7rt2zHgEE9WnLnz3rRpmm9qMMTqVGVtUHdAJydsFwHOBhoQDAFvBKUSArE43HueX8Oj4xZ\nwOZtOzCDo/dtxR0/66U2Jqm1KktQRe6+JGF5jLuvAdaYWeTFP5FsVxqLc8ebMxkxbjHbYnHyzTil\nVxtuPf1AmtZXrzyp3SpLUM0SF9z9VwmLrRCRalm/pZT/fWU6b3y1gljcKcw3fn5wB27+cU/qF1Vl\nmjaR3FXZv4RxZnaZuz+cuNLMrgC+SF9YIrlp0ZrN3PTiV4ydtwYH6hXmc+HAjtxwwn4a+UGyhpl1\nIJgPsDXBO7FD3f2elF9nd734wtEiXiYYwXxSuLo/QVvU6e6+MtUBlUe9+CTbTVy0jv99eRrTVwTz\nMTVvUMRVR3XjooGdNVaeZJzKevGZWRugjbtPMrNGwESCnDAjlXHstgTl7quAI8zsGOCAcPXr7v5B\nKoMQyVXPT1jCP9+bw/L1WwHo2KweN560Pz/qpWkvJHu5+wrC+QHdfZOZzSQYozWlCUrvQYmkWGks\nzt3vzuapzxexaVsMgIPaNeaWUw+kf6dmlRwtEj0zKwW+Slg11N2HVrBvZ+Bj4EB335jKONQaK5Ii\nq4tLuPmBgOeiAAATCElEQVSV6bw9fSWxuJNvxnH778WfTztQ7zBJtom5+4DKdjKzhsALwLWpTk6g\nBCWyx75YsIa/vDaDr5YF/z7rF+Vz/sEduOGEfamrHnmSo8yskCA5jXD3F9NxDf3rEamGshEfHvp4\nPqs2BYO3tmlSl18d051zDu6gjg+S0yyY2uJRYKa7/zNd11GCEqmCtcWl3PbGDF77cgXbYnEM6Nex\nKTf/uCe9O6h9SWqNgcB5wFdmNiVcd5O7v5HKiyhBiSRhzJxvuePNWUxfvhEH6hTk8bN+7fifU3pq\nxAepddx9DMHMFmkVSYIys4XAJmAHSTbGidS00lic+z6Yw5OfL2Ldlu1AUI135Q+6cd5hHVWNJ5Jm\nUZagjnb31RFeX6Rcs1du4rbXZzJm7mp2xJ08g8O6Nud/T+lJz7ZNog5PpNZQFZ8IQaeHJ8Yu4pEx\nC3a+VNuobgFnDejAb4/fR+PjiUQgqn91DrxnZjuAhyp6AUwk3Zas2cKtr8/gg69XsX1H8NL6AW0a\n85vj9+HY/VtHHJ1I7RZVgjrS3ZeFY/29a2az3P3jxB3M7HLgcoCiIjVCS+qUlZYe/3QBS9YFpaX6\nRfn8tF9b/nDSfur0IJIhIh/qyMxuAYrd/e8V7aOhjiQVZq/cxO1vBG1LZaWlfVo35FfHdOfU3u0i\njk4kc2TFlO/pEE50mBcOMNgAOB74c03HIbVDaSzOg6PnMWLcIlZuDF6obVingJ/2a8PvT9iP5g1V\nWhLJVFFU8bUGXgpeRKYA+I+7vxVBHJLDxsz5lrvfm83kxeuJe/DCxoFtG3PdcWpbEskWNZ6g3H0+\n0Lumryu5b3VxCXe9PZs3vlrBppJgFPHmDYr4Wb92XHPsPjSsq554ItlE/2Ilq8VicYZ/vojhny1k\n4ZotABTkGYP3acn1x+9Lr/ZNow1QRKpNCUqy0ufz1nD3e7OZsGgdO+JBh4cuLetz4RFdNMqDSI5Q\ngpKssWTdFu5+Zzbvzli5cyLApvULOeWgNvzmuH3V4UEkxyhBSUYrLolx/0dzeXHSUr4Je+EV5huD\nerTkN8ftQ9+OGkFcJFcpQUnGicfjPD1+CcPGLmTOymKcoBfe/m0acemRXfhJ33aqwhOpBZSgJGO8\nM/0bHvp4HlOWbNjZrtS2aT3OGtCeywd31Xh4IrWM/sVLpCYuWse9H8zhs3lr2BaLA9CsfiEnHrA3\nvz62B22a1os4QhGJihKU1LgZKzZw3wdz+Xj2txRv2wEEY+GddODeXHNsD/Zr0zjiCEUkEyhBSY1Y\ntGYz97w3h/dnrWLD1mDyv6KCPAZ2a8HVx/TgsG4tIo5QRJJlZo8BpwCr3P3AdF1HCUrSZsmaLdz7\n4Rzem7mKtZtLgeAl2kM6N+PSQV05/oC9I45QRKrpCeA+YHg6L6IEJSm1ZN0W7v9gLu/MWMmaMCnl\nm9GrfRMuHtiFU3u3UQ88kSzn7h+bWed0X0cJSvbYkjVbuO+jubw/YyWrd0lKQw7rxBn91C1cJMsU\nmNmEhOWhUUwsqwQl1TJ35Sb+/dE8Pvp6Feu2BG1K+WYc1K4xQw7rxJn92yspiWSvmLsPiDoIJShJ\n2rRlG3hw9Dw+mfMtG7YGQw0V5Bl9OjRlyKEd+alKSiKSQkpQsltj5nzLI2MW8MWCtWwpDbqEF+Yb\nAzo14/wjOnHKQWpTEpH0UIKS74nH47z65QpGjFvM1CXrd748W7cwjyO7t+CigV004Z9ILWdmTwNH\nAS3NbClws7s/mvLruHuqz5lyDRo08M2bN0cdRs7aUhpjxOeLeHHSMmavLGZH+HeiUZ0CDu/WgksH\ndeGQLnpPSaS2MLMt7t4g6jhUgqqlVm4s4ZFP5vPWtG9Yum4rZf9NadmwiKP33YvLBndln9aNIo1R\nRGo3JahaZPLidTw6ZgGfzl29s+edAR2b1+fEA/fmssFdaNmwbrRBioiElKByWCwWtCc9M34xU5eu\np2R70J5UkBd0B/9pv/acc3AH6mqUcBHJQPpmyjGri0t4bMxC3pr2DQvXbCactYKGdQo4et8WXHBE\nZ47ad69ogxQRSYISVA4YN38Nwz9byNh5a3ZW3QG0aVKXo/dtxSWDutKtVcPoAhQRqQYlqCxUXBLj\n6S8WMWrqcmZ9s4ntO4JiUkGecUDbxpzety3nHtpJE/yJSFbTN1iWmLpkHcPGLuLTuatZuWnbzvVN\n6xdyaOfmDDm8E4N6tIowQhGR1FKCylDFJTGeHb+YUV8uZ9aKTTtfmM0z6NaqIScc0JoLj+jMXo3V\n605EcpMSVAYZN38NI8Yt5vP5a1iVUEpqXLeAI7q14OyDO3Jcz700tJCI1ApKUBFaXVzCk58t4p3p\nK5n7bfHOtqR8M7q1asjxPVtz4cDOtFYpSURqISWoGhSLxXl92gpenLSMKUvW7RwRHKB5gyIO7tSM\nnx/SkaP2aalSkojUekpQafbl0vWMGLeYT+euZlnCkEJ1CvI4sG1jftSrDece0pEm9YsijVNEJNMo\nQaXYyo1Btd37s1Yyd9V31XZm0L5ZPQb1aMmQwzrRs22TiCMVEclsSlB7aEtpjOcnLuWNr1YwbdkG\nirft2LmtWf1C+nZsyul92/GjA9pQUKBqOxGRZClBVVEsFueN6St4ZfJyJi9Zz9rNpTu31SvMp1/H\nppx00N6c1b+Dqu1ERPaAElQl4vE4n8xZzQuTljF+4Vq+2VCysx2pMN/Yp3VDjt53L4Yc2okOLepH\nGquISC5RgirHxEXreHb8Yj6ft4al67fuHHA1z6BDs3oM7N6Ssw/pQO8OzaINVEQkhylBEQwj9Oz4\nJXw2fw2L125lR5iRDNi7SV0O6dKcM/q3Z2C3Fur+LSJSQyJJUGZ2InAPkA884u531OT1py5Zx3MT\nljJ23hoWr91CLCEhtWpUh74dmnJ6v3Yct19rdWwQESlHTXyPm7tXvlcqL2iWD8wGjgOWAuOBc9x9\nRkXHNGjQwDdv3lzta05evI7nJizhs/lrWZKQkABaNqxD3w5N+HGftuppJyICmNkWd2+wm+1V/h6v\njihKUIcAc919PoCZPQOcBqT0xgBKSmMc9Kd3vnsXCWjZqA592jfhlN5tOenANhQpIYmIVFWNfI9H\nkaDaAUsSlpcCh+66k5ldDlwOUFRUve7adYsK6Nm2MW0a1+WUXm058YC9VUISEalcgZlNSFge6u5D\nE5aT+h7f4yBSfcJUCR/GUAiq+Kp7nleuOjJlMYmI1BIxdx8QdRBRFCeWAR0SltuH60REJDvUyPd4\nFAlqPNDDzLqYWRFwNjAqgjhERKR6auR7vMar+Nw9Zma/At4m6J74mLtPr+k4RESkemrqe7zGu5lX\nx552MxcRkeRV1s28pqhLm4iIZCQlKBERyUhKUCIikpGUoEREJCNlRScJM4sDW/fgFAVALEXh5BI9\nl4rp2VRMz6Z8ufRc6rl75AWYrEhQe8rMJmTCW9GZRs+lYno2FdOzKZ+eS+pFniFFRETKowQlIiIZ\nqbYkqKGV71Ir6blUTM+mYno25dNzSbFa0QYlIiLZp7aUoEREJMsoQYmISEbK6QRlZiea2ddmNtfM\nbow6niiZ2WNmtsrMpiWsa25m75rZnPB3syhjjIKZdTCzD81shplNN7NrwvV6NmZ1zewLM5saPps/\nhetr/bMBMLN8M5tsZq+Fy3ouKZazCcrM8oF/AycBPYFzzKxntFFF6gngxF3W3Qi87+49gPfD5dom\nBvzW3XsChwFXhX9P9GxgG3CMu/cG+gAnmtlh6NmUuQaYmbCs55JiOZuggEOAue4+391LgWeA0yKO\nKTLu/jGwdpfVpwHDws/DgNNrNKgM4O4r3H1S+HkTwRdOO/Rs8EBxuFgY/jh6NphZe+Bk4JGE1bX+\nuaRaLieodsCShOWl4Tr5Tmt3XxF+/gZoHWUwUTOzzkBfYBx6NsDOaqwpwCrgXXfXswn8C7gBiCes\n03NJsVxOUFIFHrxvUGvfOTCzhsALwLXuvjFxW21+Nu6+w937AO2BQ8zswF2217pnY2anAKvcfWJF\n+9TG55IOuZyglgEdEpbbh+vkOyvNrA1A+HtVxPFEwswKCZLTCHd/MVytZ5PA3dcDHxK0Y9b2ZzMQ\nONXMFhI0HRxjZk+h55JyuZygxgM9zKyLmRUBZwOjIo4p04wCLgg/XwC8EmEskTAzAx4FZrr7PxM2\n6dmYtTKzpuHnesBxwCxq+bNx9z+4e3t370zwvfKBuw+hlj+XdMjpkSTM7EcEdcX5wGPuflvEIUXG\nzJ4GjgJaAiuBm4GXgZFAR2ARcJa779qRIqeZ2ZHAJ8BXfNeecBNBO1Rtfza9CBr78wn+MzvS3f9s\nZi2o5c+mjJkdBVzv7qfouaReTicoERHJXrlcxSciIllMCUpERDKSEpSIiGQkJSgREclISlAiIpKR\nlKBqOTPz8CXDsuUCM/u2bITm3RzX1Mx+mcT5v7efmbU1s+f3LOqd5xoUjrI9JXxPJ2cl+7wzgZld\na2b1E5bfKHufSqQqlKBkM3Bgwhf8cSQ34kZTIJkvzO/t5+7L3f2MKkdZvnOB2929j7tvLVtpZgUp\nOn8mSfZ5p50FdvfdcS2wM0G5+4/CkShEqkQJSgDeIBiZGeAc4OmyDWZ2i5ldn7A8LRxU9Q6gW1h6\nucvMGprZ+2Y2ycy+MrOykeN33a9z2ZxU4XxDj4f7Tzazo8P1F5rZi2b2Vji3zp27BmxmlwJnAX8x\nsxFmdpSZfWJmo4AZ4T5DwvmMppjZQ+EULJjZRWY2O9z2sJndF65/wszOSLhGccLn35nZeDP7MmFe\npM5mNjM8x3Qze6cs0ZtZdzN7z4K5lCaZWTczG25mpyecc0TCcypbl9RzLOd5/I8Fc5+NMbOny/7M\nwuu+ZWYTw+ezX8K9/p+ZjTWz+bvcd0X3+rWZDQemAR3M7AEzm2Dfnyvq10Bb4EMz+zBct9DMWoaf\nfxP+HZpmZtdW9hx3ucduZvZ5+FxuTfzzkRzl7vqpxT9AMdALeB6oC0whGHHitXD7LQRvypftPw3o\nHP5MS1hfADQOP7cE5gJWzn47l4HfEozwAbAfsDiM4UJgPtAkXF4EdCgn9ieAM8LPRxGUBruEy/sD\nrwKF4fL9wPlAm/A6rYAi4FPgvl3PV/Zswt/HA0PD+8kDXgMGh/cSA/qE+40EhoSfxwE/CT/XJShR\n/AB4OVzXBFgAFOxyT0k9x12OOTj8c6sLNALmlP2ZEcxL1CP8fCjBsDxl9/pceD89Caamqexe48Bh\nCddtHv7OBz4CeoXLC4GWCfstDO+lP8GIHQ2AhsB0gtHjK3yOu9zna8A54ecry/589JO7P7lYFSJV\n5O5fhqWicwhKU9VhwF/NbDDBF1k7Kp9u4Ejg3jCGWWa2CNgn3Pa+u28AMLMZQCe+P31Keb5w9wXh\n52MJvhDHmxlAPYLBOw8FPnL3b8NzP5twzYocH/5MDpcbAj0IEt0Cd58Srp8IdDazRkA7d38pvLeS\ncPtoM7vfzFoBPwNecPfYLteqznMcCLwSXqfEzF4N760hcATwXPgMAOokHPeyu8eBGWZWdo3d3esi\nd/884fizzOxygqTahiDRfbmbOI8EXnL3zWF8LwKDCMaw+/+eYznHH853cyz9B/j7bq4lOUAJSsqM\nIvgHfxTQImF9jO9XBdet4PhzCUol/d19uwUjPVe0bzK2JXzeQXJ/VzcnfDZgmLv/IXGHxCq2cuy8\nVwvaWIoSznW7uz+0y7k6lxNnZZ01hgNDCAYZvaic7al8jnnAeg+myyhPYuyW8Luie92csNwFuB44\n2N3XmdkTexDnrrEk8xylFlAblJR5DPiTu3+1y/qFQD8AM+sHdAnXbyKoTirThGCOnO1hW1KnCvZL\n9AnBFzJmtg/BIJtf79lt7PQ+cIaZ7RWev7mZdSKoevuBmbWwYJqNMxOOWUhQ6gI4lWAGWYC3gYvD\nEglm1q7svOXxYGbepWXJ0Mzq2He92p4g6ESAu88o5/DqPMdPgR9b0KbXEDglPP9GYIGZnRnGYWbW\nu6K4q3ivjQkS1oaw9HVSwraKYv0EON3M6ptZA+An4bpkfU5Q8oQgwUuOU4ISANx9qbv/XzmbXgCa\nm9l04FfA7HD/NcCnYWP3XcAIYICZfUXQ1jOrgv0S3Q/khcc8C1zo7ttIgfDL/4/AO2b2JfAu0MaD\nGU9vAT4j+GKfmXDYwwTJaypBddLm8FzvEFQpfRbG+jwVJ4sy5wG/Dq89Ftg7PNfK8JqPV3BclZ+j\nu48nKAF/CbxJ0M6zIdx8LnBJeE/TCaYlr1Cy9+ruUwmqAWeF+3+asHko8FZZJ4mEYyYRJOgvCP6j\n8Ii7TyZ51wK/CZ9p94R7lByl0cylVjOzC4EB7v6rGrpefYIE0q+sjS1F523o7sXh+T8GLg8TQs4I\n722ru7uZnU3QYWK3CVeym9qgRGqImf2QYHLEu1OZnEJDzawnQTvQsFxLTqH+wH0W9PhYD1wccTyS\nZipBiYhIRlIblIiIZCQlKBERyUhKUCIikpGUoEREJCMpQYmISEb6fyOpNzLhM/O1AAAAAElFTkSu\nQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x1b51cd26390>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "years = lambda generations: generations * 1.2 / 365\n",
    "\n",
    "fig, ax1 = plt.subplots()\n",
    "x = np.linspace(1.0, 0.55, 50)\n",
    "gs = estimate_generations(x, 5*10**(-4), 5*10**(-4))\n",
    "x_axis = 100 * (1 - x)\n",
    "\n",
    "ax1.plot(x_axis, gs / 10**2)\n",
    "ax1.set_xlabel('Mutation frequency at generation g')\n",
    "ax1.set_ylabel('Generations (100)')\n",
    "ax1.set_title('Time since infection')\n",
    "ax2 = ax1.twinx()\n",
    "ax2.plot(x_axis, years(gs))\n",
    "ax2.set_ylabel('Years')\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Identifying the sets $S_{RD}$ and $S_{DCC}$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def s0_sequence(pileup: List[dict], threshold=0.8):\n",
    "    \"\"\" Returns the sequence of nucleotides based on major threshold.\n",
    "    \"\"\"\n",
    "    freqs_dict = lambda d: {key: value for key, value in d.items() if 'freq' in key}\n",
    "    major_base = lambda freqs_dict: max(freqs_dict.items(), key=lambda d: d[1])[0][0]\n",
    "    s0 = ''.join([major_base(freqs_dict(d)) for d in pileup])\n",
    "    return s0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def codonify_orf(sequence):\n",
    "    \"\"\" Returns sequence as list of codons. \"\"\"\n",
    "    s = s0.find('ATG')\n",
    "    if s == -1:\n",
    "        raise ValueError('srd_filter cannot find start codon in sequence:\\n%s' % s0)\n",
    "    firsts, seconds, thirds = s0[s::3], s0[(s+1)::3], s0[(s+2)::3]\n",
    "    codons = [first + second + third\n",
    "              for first, second, third in zip(firsts, seconds, thirds)]\n",
    "    return codons, s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def find_srd_locs(pileup: List[dict], s0=None):\n",
    "    \"\"\" Returns those genome locations which are in the set of death reduced locations.\n",
    "    \n",
    "    :param pileup: list of dictionaries of locations (must include X_freq keys for each base X).\n",
    "    :return: List of integers with 1-based location of final Ts and Cs in TAT and TAC codons.\n",
    "    \"\"\"\n",
    "    s0 = s0 or s0_sequence(pileup)\n",
    "    int_five_percent_codons = len(s0) // 20\n",
    "    codons, orf_offset = codonify_orf(s0)\n",
    "    codons = codons[:-int_five_percent_codons] if int_five_percent_codons > 0 else codons\n",
    "    if len(codons) == 0:\n",
    "        raise ValueError('Too short sequence given: %s' % s0)\n",
    "    srd_codon_nums = [i + 1 for i, codon in enumerate(codons)\n",
    "                      if codon == 'TAT' or codon == 'TAC']\n",
    "    srd_locs = [codon_num*3 + orf_offset for codon_num in srd_codon_nums]\n",
    "    return srd_locs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def find_sdcc_locs(pileup: List[dict], s0=None):\n",
    "    \"\"\" Returns those genome locations which are in the set of death codon cytosines.\n",
    "    \n",
    "    :param pileup: list of dictionaries of locations (must include X_freq keys for each base X).\n",
    "    :return: List of integers with 1-based location of first Cs in CGA, CAA and CAG codons.\n",
    "    \"\"\"\n",
    "    s0 = s0 or s0_sequence(pileup)\n",
    "    int_five_percent_codons = len(s0) // 20\n",
    "    codons, orf_offset = codonify_orf(s0)\n",
    "    codons = codons[:-int_five_percent_codons] if int_five_percent_codons > 0 else codons\n",
    "    if len(codons) == 0:\n",
    "        raise ValueError('Too short sequence given: %s' % s0)\n",
    "    sdcc_codon_nums = [i + 1 for i, codon in enumerate(codons)\n",
    "                       if codon in ['CGA', 'CAA', 'CAG']]\n",
    "    sdcc_locs = [codon_num*3 - 2 + orf_offset for codon_num in sdcc_codon_nums]\n",
    "    return sdcc_locs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Count TAC to TAT ratio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calc_tac_tat_ratio(sequence: str) -> float:\n",
    "    return sequence.count('TAC') / sequence.count('TAT')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Estimate substitution rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def estimate_qtc(pileup: List[dict], s0=None, sdcc_locations=None) -> (float, float, float):\n",
    "    sdcc_locs = sdcc_locations or find_sdcc_locs(pileup, s0)\n",
    "    sdcc = [pileup[i - 1] for i in sdcc_locs]\n",
    "    estimates = [d['T_freq'] for d in sdcc]\n",
    "    return np.mean(estimates), min(estimates), max(estimates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def estimate_qct(pileup: List[dict], tac_tat_ratio: float=None,\n",
    "                 qtc: float=None, s0: str=None) -> (float, float, float):\n",
    "    s0 = s0 or s0_sequence(pileup)\n",
    "    r = tac_tat_ratio or calc_tac_tat_ratio(s0)\n",
    "    qtc_mean, qtc_min, qtc_max = estimate_qtc(pileup, s0=s0) if qtc is None else (qtc, qtc, qtc)\n",
    "    qct_mean, qct_min, qct_max = r * qtc_mean, r * qtc_min, r * qtc_max\n",
    "    return qct_mean, qct_min, qct_max"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Putting them together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimate_patient_tsi(pileup: List[dict]) -> (float, float, float):\n",
    "    \"\"\" Returns tuple of mean, min and max estimates of time since infection (generations). \"\"\"\n",
    "    s0 = s0_sequence(pileup)\n",
    "    qtc_mean, qtc_min, qtc_max = estimate_qtc(pileup, s0=s0)\n",
    "    qct_mean, qct_min, qct_max = estimate_qct(pileup, s0=s0, qtc=qtc_mean)\n",
    "    \n",
    "    srd = [pileup[i - 1] for i in find_srd_locs(pileup)]\n",
    "    srd_s0 = s0_sequence(srd)\n",
    "    p0 = srd_s0.count('C') / (srd_s0.count('T') + srd_s0.count('C'))\n",
    "    pgs = [1 - min(d['C_freq'], d['T_freq']) for d in srd]\n",
    "\n",
    "    try:\n",
    "        g_estimates = estimate_generations(pgs, qtc_mean, qct_mean, p0)\n",
    "    except Exception as e:\n",
    "        print('p0: ', p0)\n",
    "        print('qct: ', qct_mean)\n",
    "        print('qtc: ', qtc_mean)\n",
    "        print('pgs: ', pgs)\n",
    "        raise e\n",
    "    \n",
    "    return np.mean(g_estimates), min(g_estimates), max(g_estimates)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Running tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dA = {'A_freq': 0.97, 'C_freq': 0.01, 'T_freq': 0.01, 'G_freq': 0.01, 'reference': 'C', 'depth': 1000}\n",
    "dT = {'A_freq': 0.01, 'C_freq': 0.01, 'T_freq': 0.97, 'G_freq': 0.01, 'reference': 'C', 'depth': 1000}\n",
    "dG = {'A_freq': 0.01, 'C_freq': 0.01, 'T_freq': 0.01, 'G_freq': 0.97, 'reference': 'C', 'depth': 1000}\n",
    "dC = {'A_freq': 0.01, 'C_freq': 0.97, 'T_freq': 0.01, 'G_freq': 0.01, 'reference': 'C', 'depth': 1000}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGTATTACCAACAGTAC'"
      ]
     },
     "execution_count": 375,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pile = [dA, dT, dG]\n",
    "pile = pile + [dT, dA, dT] + [dT, dA, dC] + [dC, dA, dA] + [dC, dA, dG] + [dT, dA, dC]\n",
    "s0_sequence(pile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGTATTACCAACAGTAC'"
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s0 = s0_sequence(pile)\n",
    "s0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 377,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[6, 9, 18]\n",
      "[{'A_freq': 0.01, 'C_freq': 0.01, 'T_freq': 0.97, 'G_freq': 0.01, 'reference': 'C', 'depth': 1000}, {'A_freq': 0.01, 'C_freq': 0.97, 'T_freq': 0.01, 'G_freq': 0.01, 'reference': 'C', 'depth': 1000}, {'A_freq': 0.01, 'C_freq': 0.97, 'T_freq': 0.01, 'G_freq': 0.01, 'reference': 'C', 'depth': 1000}]\n",
      "TCC\n"
     ]
    }
   ],
   "source": [
    "print(find_srd_locs(pile))\n",
    "srd = [pile[i - 1] for i in find_srd_locs(pile)]\n",
    "print(srd)\n",
    "srd_s0 = s0_sequence(srd)\n",
    "print(srd_s0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 378,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[10, 13]"
      ]
     },
     "execution_count": 378,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_sdcc_locs(pile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.01, 0.01, 0.01)"
      ]
     },
     "execution_count": 379,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "estimate_qtc(pile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.02, 0.02, 0.02)"
      ]
     },
     "execution_count": 380,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "estimate_qct(pile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 381,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.50378792700160613, 0.50378792700160613, 0.50378792700160613)"
      ]
     },
     "execution_count": 381,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "estimate_patient_tsi(pile)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
